enterprise_gateway/services/kernelspecs/handlers.py

```diff
@@ -1,17 +1,175 @@
+            name,
+            fname
+        )
+    return d
+
 
 # Extends the default handlers from the notebook package with token auth, CORS
 # and JSON errors.
 default_handlers = []
-for path, cls in notebook_handlers.default_handlers:
-    # Everything should have CORS and token auth
+
+
+def exists(obj, chain):
+    _key = chain.pop(0)
+    if _key in obj:
+        return exists(obj[_key], chain) if chain else obj[_key]
+
+
+def apply_user_filter(kernelspec_model, kernel_user=None):
+    if kernel_user:
+        # Check unauthorized list
+        if exists(kernelspec_model, ['spec', 'metadata', 'process_proxy', 'config', 'unauthorized_users']):
+            # Check if kernel_user in kernelspec_model
+            unauthorized_list = kernelspec_model['spec']['metadata']['process_proxy']['config']['unauthorized_users']
+            if kernel_user in unauthorized_list:
+                return None
+        # Now we have check that unauthorized list doesnt exist
+        if exists(kernelspec_model, ['spec', 'metadata', 'process_proxy', 'config', 'authorized_users']):
+            authorized_list = kernelspec_model['spec']['metadata']['process_proxy']['config']['authorized_users']
+            if authorized_list and kernel_user not in authorized_list:
+                return None
+    return kernelspec_model
+
+
+class ModifyKernelSpecHandler(APIHandler):
+    @web.authenticated
+    @gen.coroutine
+    def get(self, kernel_user=None):
+        ksm = self.kernel_spec_manager
+        km = self.kernel_manager
+        model = {}
+        model['default'] = km.default_kernel_name
+        model['kernelspecs'] = specs = {}
+        kspecs = yield maybe_future(ksm.get_all_specs())
+        if kernel_user:
+            self.log.info("Searching kernels for user '%s' " % kernel_user)
+        else:
+            self.log.info("No user. All kernels given")
+
+        for kernel_name, kernel_info in kspecs.items():
+            try:
+                if is_kernelspec_model(kernel_info):
+                    d = kernel_info
+                else:
+                    d = kernelspec_model(self, kernel_name, kernel_info['spec'], kernel_info['resource_dir'])
+
+                d = apply_user_filter(d, kernel_user)
+                if d is not None:
+                    self.log.info("Find kernel '%s'", d['name'])
+                    specs[kernel_name] = d
+
+            except Exception:
+                self.log.error("Failed to load kernel spec: '%s'", kernel_name)
+                continue
+        self.set_header("Content-Type", 'application/json')
+        self.finish(json.dumps(model))
+
+
+kernel_name_regex = r"(?P<kernel_name>[\w\.\-%]+)"
+kernel_user_regex = r"(?P<kernel_user>[\w%]+)"
+
+default_handlers_notebooks = [
+    (r"/api/kernelspecs", ModifyKernelSpecHandler),
+    (r"/api/kernelspecs/user/%s" % kernel_user_regex, ModifyKernelSpecHandler),
+    (r"/api/kernelspecs/%s" % kernel_name_regex, notebook_handlers.KernelSpecHandler)
```

Comment:
Although this is a supported endpoint (`/api/kernelspecs/<kernel-name>`), I'm not sure there's anything that calls it from Notebook.  However, it is supported.  Do we need a similar handler that allows a user to be specified?

Reply:
I need to take a look to `/api/kernelspecs/<kernel-name` function.

Reply:
`/api/kernelspecs/<kernel-name` returns a spec model of available kernels with given kernel name.

I my opinion, I think we dont need to have an specified handler because I guess that this `kernel-name` will have this user authorized.

Is this call made after calling /api/kernelspecs?

Reply:
>Is this call made after calling /api/kernelspecs?

I'm not sure either notebook front-ends (notebook, lab) call this endpoint.  I've set breakpoints in this handler and they are not hit, iirc.

>I my opinion, I think we dont need to have an specified handler because I guess that this kernel-name will have this user authorized.

I don't think that's a valid assumption. If I call /api/kernelspecs/foo, yet I am not authorized to run kernels of this type, should I be able to access this kernelspec - if there was a means to enforce authorization on this endpoint like you're adding on the other?

This PR should also add the /user endpoint definition in the swagger files in https://github.com/jupyter/enterprise_gateway/tree/master/enterprise_gateway/services/api.  I can help you with that although we should probably wait to see how this discussion turns out (since we may need to add another handler).